### PR TITLE
Update E0088 to new format

### LINKED ISSUE #35226: E0088 needs to be updated to new format
From: src/test/compile-fail/E0088.rs

Error E0088 needs to be updated with a span_label, taking it from:

```
error[E0088]: too many lifetime parameters provided: expected 0 parameters, found 1 parameter
  --> src/test/compile-fail/E0088.rs:14:9
   |
14 |     f::<'static>(); //~ ERROR E0088
   |         ^^^^^^^
```

To:

```
error[E0088]: too many lifetime parameters provided: expected 0 parameters, found 1 parameter
  --> src/test/compile-fail/E0088.rs:14:9
   |
14 |     f::<'static>(); //~ ERROR E0088
   |         ^^^^^^^ 0 lifetime parameters expected
```

Bonus: we may want to detect the 0 case and say:

```
error[E0088]: too many lifetime parameters provided: expected 0 parameters, found 1 parameter
  --> src/test/compile-fail/E0088.rs:14:9
   |
14 |     f::<'static>(); //~ ERROR E0088
   |         ^^^^^^^ unexpected lifetime parameter
```

### PR DESCRIPTION
**Pull request by yossi-k**
Fixes #35226.
Part of #35233.

r? @jonathandturner

### CODE REVIEW

File: src/librustc_typeck/check/mod.rs [hunk 1]
```diff
@@ -4364,11 +4364,17 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
-                      count(lifetime_defs.len()),
-                      count(lifetimes.len()));
+            let mut err = struct_span_err!(self.tcx.sess, span, E0088,
+                                           "too many lifetime parameters provided: \
+                                            expected {}, found {}",
+                                           count(lifetime_defs.len()),
+                                           count(lifetimes.len()));
+            match lifetime_defs.len() {
+                0 => err.span_label(span, &format!("unexpected lifetime parameter")),
+                _ => err.span_label(span,  &format!("{} lifetime parameters expected",
```
Comment (jonathandturner) [inline src/librustc_typeck/check/mod.rs]: Now that we handle multiple cases, can you add to the unit test to check for these new cases?

**Comment by jonathandturner** [comment]
Thanks for the PR!

Looks like we'll need to test for some of your new cases.  After that I think we're probably good to go.

**Comment by yossi-k** [comment]
@jonathandturner I'm not much of a rust expert, but does the second case make sense? I can't think of a code to trigger this, perhaps you can point me in some useful direction.

**Comment by jonathandturner** [comment]
You could try an example that tests against two different functions, one with a lifetime and one without.  Something like:

```
fn f() {}

fn g<'a>() {}

fn main() {
    f::<'static>(); //~ ERROR E0088
    //~^ unexpected lifetime parameter

    g::<'static, 'static>();
}
```

**Comment by yossi-k** [comment]
@jonathandturner That was also my first guess, but ```g``` doesn't expect any lifetimes at all, we need something that expects more than zero.

**Comment by jonathandturner** [comment]
@yossi-k - looks like you're right.  When I try it with functions, the lifetime parameters seem to be ignored.  If I try it with traits, I just get a different error message:

```
fn f() {}

fn g<'a>(x: i32) -> &'a str {}

trait Foo<'a> { }

fn main() {
    f::<'static>(); //~ ERROR E0088
    //~^ unexpected lifetime parameter

    g::<'static, 'static>(4);

    let x: Foo<'static, 'static>;
}
```

It may not be possible to invoke the code path.  If that's the case, it may make sense to always say "unexpected lifetime parameters" in the label

**Comment by yossi-k** [comment]
@jonathandturner So considering the fact that it doesn't seem possible to actually provide a lifetime parameter to a function, and providing lifetimes to other things (such as traits) is covered by a whole different set of errors (e.g. E0107), is [this code segment](https://github.com/yossi-k/rust/blob/7f6ad7fbaa565fdad2eabbfe5c7a7e7d423741b6/src/librustc_typeck/check/mod.rs#L4363-L4384) actually needed?

**Comment by jonathandturner** [comment]
@yossi-k - you could try taking it out and running the unit test suite again and see what happens.  If all of the errors are still caught by E0107, you might be right.  And it would help us simplify things.

**Comment by yossi-k** [comment]
@jonathandturner All tests seem to pass, so I'm guessing we're okay. 
Also I merged the Spans of all provided lifetime parameters.

**Comment by jonathandturner** [comment]
@yossi-k - unfortunately, you can't safely merge multiple spans together, since spans can come from macros and might give you a broken span as a result.

**Comment by yossi-k** [comment]
@jonathandturner What should I do in this case? Only reference one lifetime parameter, perhaps each individually or maybe just the entire function call?

**Comment by jonathandturner** [comment]
@yossi-k - I just added a way to safely merge spans to CodeMap.  It's a new method called `merge_spans` that will merge the spans if it's safe to do so.

**Comment by jonathandturner** [comment]
@yossi-k - after working some with @GuillaumeGomez, I think a better solution than merging spans would be to instead point at the argument when it's unexpected.  You can see an example on [E0050](https://github.com/rust-lang/rust/issues/35211#issuecomment-249641000)

**Comment by GuillaumeGomez** [comment]
@jonathandturner: It hasn't been updated yet. :p

**Comment by jonathandturner** [comment]
@GuillaumeGomez - moved it from a comment to the issue text.  Should be updated now

**Comment by alexcrichton** [comment]
Closing due to inactivity, but feel free to resubmit with a rebase!

**Comment by eddyb** [comment]
@jonathandturner You should've asked a rustc developer 😭. This (through #37835) introduced a regression which I only found because I was rewriting lifetime elision and I had to update this check.

Thankfully it's only propagated to beta so we can just backport a fix, but this is scary that it just happened.
Well, not *that* scary, what will happen is inference variables will be created for the missing lifetimes.

What's confusing here is the distinction between early-bound and late-bound `fn` lifetime parameters.
Early-bound lifetime parameters work like they do in type definitions, they get passed in the path and applied to the type to get the type of the path, and all uses of the resulting value get the same lifetime.
They *only* end up existing when the syntactic lifetime is used in a bound (e.g. `'b: 'a` and `T: Trait<'a>`).

Late-bound lifetimes, explicitly expressed by `for<'a>` when nested in types, are the most common case, and they allow a signature to stay generic (and work with e.g. `Fn(&T)` aka `for<'a> Fn(&'a T)`).

For some historical reason or another, we ended up mapping lifetimes passed to a path to only the early-bound lifetimes and have no mechanism for "downgrading" late- to early-bound for paths with lifetimes.
We may want to change to something like this, I'm not sure (explicit lifetimes to value paths are rare IME).

To wrap it up, this is the test I used (errors on 1.14 stable, incorrectly passes on beta/nightly):
```rust
fn foo<'a: 'b, 'b: 'a>() {}
fn main() {
    foo::<'static>();
}
```

cc @rust-lang/compiler @rust-lang/lang

**Comment by arielb1** [comment]
> For some historical reason or another, we ended up mapping lifetimes passed to a path to only the early-bound lifetimes and have no mechanism for "downgrading" late- to early-bound for paths with lifetimes.

That's just an implementation artifact, methinks.

**Comment by jonathandturner** [comment]
@eddyb - you're commenting on a closed PR, I assume you meant the merged one?

Glad we could catch it in time.  Looking at it now, definitely an oversight on my part.  Clearly they changed the logic rather than just changing the message.

**Comment by jonathandturner** [comment]
@eddyb - potential fix at https://github.com/rust-lang/rust/pull/38859

**Pull request closed without merging**